pkg/didcomm/crypto/authcrypt/encrypt.go

```diff
@@ -0,0 +1,288 @@
+
+	recipients, err := c.encodeRecipients(symKey, senderEphemeralKey)
+	if err != nil {
+		return "", err
+	}
+
+	json, err := c.buildJWE(headers, recipients, aadEncoded, nonceEncoded, tagEncoded, cipherTextEncoded)
+	if err != nil {
+		return "", err
+	}
+
+	return string(json), nil
+}
+
+func (c *Crypter) buildJWE(headers jweHeaders, recipients []Recipient, aad string, iv string, tag string, ciphertext string) ([]byte, error) {
+
+	jwe := JweEnvelope{
+		Protected:  headers,
+		Recipients: recipients,
+		Aad:        aad,
+		Iv:         iv,
+		Tag:        tag,
+		CipherText: ciphertext,
+	}
+
+	jweBytes, err := json.Marshal(jwe)
+	if err != nil {
+		return nil, err
+	}
+
+	return jweBytes, nil
+}
+
+func (c *Crypter) buildAAD() []byte {
+	var kids []string
+	for _, r := range c.Recipients {
+		recipientKey := r.(*ecdsa.PublicKey)
+		kids = append(kids, base58.Encode(elliptic.Marshal(recipientKey, recipientKey.X, recipientKey.Y)))
+	}
+	sort.Strings(kids)
+	sha := sha256.Sum256([]byte(strings.Join(kids, ".")))
+	return sha[:]
+}
+
+func (c *Crypter) encodeRecipients(symKey []byte, senderEphemeralKey *ecdsa.PrivateKey) ([]Recipient, error) {
+	var encodedRecipients []Recipient
+	for _, e := range c.Recipients {
+		rec, err := c.encodeRecipient(symKey, senderEphemeralKey, e.(*ecdsa.PublicKey))
+		if err != nil {
+			return nil, err
+		}
+		encodedRecipients = append(encodedRecipients, *rec)
+	}
+	return encodedRecipients, nil
+}
+
+func (c *Crypter) encodeRecipient(symKey []byte, senderEphemeralKey *ecdsa.PrivateKey, recipientKey *ecdsa.PublicKey) (*Recipient, error) {
+	apu := make([]byte, 64)
+	_, err := RandReader.Read(apu)
+	if err != nil {
+		return nil, err
+	}
+
+	apuEncoded := make([]byte, base64.URLEncoding.EncodedLen(len(apu)))
+	base64.URLEncoding.Encode(apuEncoded, apu)
+
+	// create authentication key for symmetric shared key
+	kek := josecipher.DeriveECDHES(c.Alg, apuEncoded, nil, senderEphemeralKey, recipientKey, chacha20poly1305.KeySize)
+
+	var cipher cipher.AEAD
+	switch c.NonceSize {
+	case chacha20poly1305.NonceSize:
+		cipher, err = chacha20poly1305.New(kek)
+	case chacha20poly1305.NonceSizeX:
+		cipher, err = chacha20poly1305.NewX(kek)
+	default:
+		return nil, errors.New("Bad nonce size, can't encrypt")
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	nonce := make([]byte, c.NonceSize)
+	_, err = RandReader.Read(nonce)
+	if err != nil {
+		return nil, err
+	}
+
+	// encrypt symmetric shared key using authentication key
+	kekOutput := cipher.Seal(nil, nonce, symKey, nil)
+
+	tag := kekOutput[len(kekOutput)-poly1305.TagSize:]
+	symKeyCipher := kekOutput[0 : len(kekOutput)-poly1305.TagSize]
+
+	return c.buildRecipient(symKeyCipher, apu, nonce, tag, recipientKey)
+}
+
+func (c *Crypter) buildRecipient(symKeyCipher []byte, apu []byte, nonce []byte, tag []byte, recipientKey *ecdsa.PublicKey) (*Recipient, error) {
+
+	/*senderKey := c.Sender.(*ecdsa.PrivateKey).PublicKey
```

Comment:
Delete all this commented-out code

Reply:
oid value is temporary and will need to be changed.. this comment is temporary till the value is properly encrypted.

Reply:
found solution for encryption oid: `sodium_crypto_box_seal` is actually nacl's `box.Seal()` (using ed25519 + Salsa20 + Poly1035) so will replace oid to call the x/crypto/nacl package instead..

Reply:
removed